### PR TITLE
Improve HTTP Status codes

### PR DESCRIPTION
**Pull request by elsewhat**
## Overview
Improve HTTP status codes section:   
- Added examples of commonly used status codes
- Added reference to official list of possible http status codes

Used https://equinor.github.io/maintenance-api-docs/#section/Modelling-of-endpoints/HTTP-Response-codes as input to the content

### CODE REVIEW

**Review by JanFredrikD — APPROVED** [review]

File: docs/rest_guidelines.md [hunk 1]
```diff
@@ -121,8 +121,22 @@ The status codes are grouped in five classes:
 * 4xx client error – the request contains bad syntax or cannot be fulfilled
 * 5xx server error – the server failed to fulfill an apparently valid request
 
-Links to status code lists:
+In total there are over 60 possible HTTP Status codes, but APIs should focus on using a limited set in a consistent manner.
+
+The most important HTTP Status codes are:
```
Comment (oyron) [inline docs/rest_guidelines.md:126]: Maybe use _commonly used_ instead of _important_?

Comment (elsewhat) [reply docs/rest_guidelines.md]: Agree, push a new commit for this

**Comment by oyron** [comment]
I think the list of status codes looks good. What do you think about including 202 - Accepted in the list? Are you using that maybe after creation of longer running jobs that are processed asynchronously?

**Comment by elsewhat** [comment]
> I think the list of status codes looks good. What do you think about including 202 - Accepted in the list? Are you using that maybe after creation of longer running jobs that are processed asynchronously?

Could be on the list. Haven't used it much myself so not sure how widely adopted it is.


From https://datatracker.ietf.org/doc/draft-ietf-httpbis-semantics/

```   
15.3.3.  202 Accepted

   The _202 (Accepted)_ status code indicates that the request has been
   accepted for processing, but the processing has not been completed.
   The request might or might not eventually be acted upon, as it might
   be disallowed when processing actually takes place.  There is no
   facility in HTTP for re-sending a status code from an asynchronous
   operation.

   The 202 response is intentionally noncommittal.  Its purpose is to
   allow a server to accept a request for some other process (perhaps a
   batch-oriented process that is only run once per day) without
   requiring that the user agent's connection to the server persist
   until the process is completed.  The representation sent with this
   response ought to describe the request's current status and point to
   (or embed) a status monitor that can provide the user with an
   estimate of when the request will be fulfilled.
```

**Comment by oyron** [comment]
We could leave it out for now. And then add it at a later stage if we see that it belongs in the list.

**Comment by fmell** [comment]
This looks good!

I think it would be nice to include a redirect status code, however I am not sure how commonly this is used. We have used this to redirect the client to an Azure Storage blob uri for downloading files, thus avoiding large files having to be streamed through our API. I have also seen it used to redirect clients to a new endpoint if a resource has been renamed. For example if a GitHub issue is moved from one repository to another where you have access, the API would redirect you to the new issue URL instead of returning a 404.

File: docs/rest_guidelines.md [hunk 2]
```diff
@@ -121,8 +121,22 @@ The status codes are grouped in five classes:
 * 4xx client error – the request contains bad syntax or cannot be fulfilled
 * 5xx server error – the server failed to fulfill an apparently valid request
 
-Links to status code lists:
+In total there are over 60 possible HTTP Status codes, but APIs should focus on using a limited set in a consistent manner.
+
+Commonly used HTTP Status codes are: 
+* 200 OK - The request completed successfully
+* 201 Created - The request completed successfully and one or more new resources were created
+* 204 No Content - The request completed successfully and there is no additional content in the response (commonly used for HTTP PATCH)
```
Comment (keaaa) [inline docs/rest_guidelines.md:129]: should the most used 30x be included? e.g. 301 and 304?

Comment (elsewhat) [reply docs/rest_guidelines.md:129]: Could be a good candidate. 

Attempted to find some more information on how common the usage of 301 and 302s are for APIs 
image
Found a couple of experts (involved in writing some of these standards) recommending HTTP 307 for temporary redirects and 308 for permanent redirects.

The main reason for this is that if you reply HTTP 301/302 for a HTTP POST request, the browser/libraries follow up with a HTTP GET for the new url. Thereby, it switches the HTTP verb and any body payload is lost. 307/308 should not have that limitation.

So perhaps 307/308 are better candidates?

Comment (keaaa) [reply docs/rest_guidelines.md:129]: Yeah, 307/308 seems to be more correct!

I think 304 is also quite often used if result is cached some time

**Review by keaaa — APPROVED** [review]

**Pull request merged**